stage #2/module#5 Engineering Best-Practices/Readme.md

```diff
@@ -0,0 +1,1275 @@
+\system performance over time. If not dealt with, the application will 
+eventually exhaust its resources, finally terminating with a 
+fatal java.lang.OutOfMemoryError.
+There are two different types of objects that reside in Heap memory,
+referenced and unreferenced. Referenced objects are those that still 
+have active references within the application, whereas unreferenced 
+objects don't have any active references.
+The garbage collector removes unreferenced objects periodically, but 
+it never collects the objects that are still being referenced.
+In any application, memory leaks can occur for numerous reasons.
+In this section, we'll discuss the most common ones.
+
+#### _Antipattern example:_
+```java
+public class StaticTest {
+  public static List<Double> list = new ArrayList<>();
+
+  public void populateList() {
+    for (int i = 0; i < 10000000; i++) {
+      list.add(Math.random());
+    }
+    Log.info("Debug Point 2");
+  }
+  
+  public static void main(String[] args) {
+    Log.info("Debug Point 1");
+    new StaticTest().populateList();
+    Log.info("Debug Point 3");
+  }
+}
+```
+If we analyze the Heap memory during this program execution, then we'll
+see that between debug points 1 and 2, the heap memory 
+increased as expected.
+
+But when we leave the populateList() method at the debug point 
+3, the heap memory isn't yet garbage collected
+However, if we just drop the keyword static in line number 2 of the above program, 
+then it'll bring a drastic change to the memory usage
+The first part until the debug point is almost the same as what we obtained in the case of static. But this time, after we leave the populateList() method, all the memory of the list is garbage collected because we don't have any reference to it.
+
+So we need to pay very close attention to our usage of static variables. If collections or large objects are declared as static, then they remain in the memory throughout the lifetime of the application, thus blocking vital memory that could otherwise be used elsewhere.
+
+>How to Prevent It?
+ - Minimize the use of static variables. 
+ - When using singletons, rely upon an implementation 
+that lazily loads the object, instead of eagerly loading.
+
+2. Through Unclosed Resources
+
+Whenever we make a new connection or open a stream, the JVM allocates memory for these resources. A few examples of this include database connections, input streams, and session objects.
+Forgetting to close these resources can block the memory, thus keeping them out of the reach of the GC. This can even happen in case of an exception that prevents the program execution from reaching the statement that's handling the code to close these resources.
+In either case, the open connection left from the resources consumes memory, and if we don't deal with them, they can deteriorate performance, and even result in an OutOfMemoryError.
+
+>How to Prevent It?
+
+- Always use finally block to close resources.
+
```

Comment:
Maybe add try-catch with resources in "How to prevent It", below you refer to it.